server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -96,7 +97,13 @@ protected ShardIterator shards(ClusterState state, InternalRequest request) {
         if (iterator == null) {
             return null;
         }
-        return new PlainShardIterator(iterator.shardId(), iterator.getShardRoutings().stream().filter(ShardRouting::isSearchable).toList());
+        return new PlainShardIterator(iterator.shardId(), iterator.getShardRoutings().stream().filter((shardRouting -> {
+            if (INDEX_FAST_REFRESH_SETTING.get(state.metadata().index(shardRouting.index()).getSettings())) {
+                return shardRouting.isPromotableToPrimary();
+            } else {
+                return shardRouting.isSearchable();
+            }
```

Comment:
nit: This filtering of the shards based on the fast_refresh setting could probably be a static method in `OperationRouting` and reused.

Reply:
When I change also how mgets get routed to search shards first, we'd need similar changes also in `TransportShardMultiGetAction`.

Reply:
Good idea, I refactored it to `OperationRouting.canSearchShard()`.

Indeed, once you do the mget changes, we'd need to do that. But good thing is that whoever merges first will cause the other to block. If my stateless PR goes through first, you'll not be able to pass the IT I introduce in that PR. If your mget PR gets merged first, my IT will fail and I will re-adapt my PR. So no need to worry.